coordinator/src/services/translation/response/spine-response-handler.ts

```diff
@@ -0,0 +1,187 @@
+    return sendMessagePayload.acknowledgement._attributes.typeCode
+  }
+
+  private extractRejectionCodes(sendMessagePayload: hl7V3.SendMessagePayload<T>) {
+    const acknowledgementDetails = sendMessagePayload.acknowledgement.acknowledgementDetail ?? []
+    return toArray(acknowledgementDetails).map(acknowledgementDetail => acknowledgementDetail.code)
+  }
+
+  private extractErrorCodes(sendMessagePayload: hl7V3.SendMessagePayload<T>) {
+    const reasons = sendMessagePayload.ControlActEvent.reason ?? []
+    return toArray(reasons).map(reason => reason.justifyingDetectedIssueEvent.code)
+  }
+
+  static createSuccessResponse(): TranslatedSpineResponse {
+    return {
+      statusCode: 200,
+      fhirResponse: fhir.createOperationOutcome([{
+        code: "informational",
+        severity: "information"
+      }])
+    }
+  }
+
+  static createBadRequestResponse(issues: Array<fhir.OperationOutcomeIssue>): TranslatedSpineResponse {
+    return {
+      statusCode: 400,
+      fhirResponse: fhir.createOperationOutcome(issues)
+    }
+  }
+
+  static createServerErrorResponse(): TranslatedSpineResponse {
+    return {
+      statusCode: 500,
+      fhirResponse: fhir.createOperationOutcome([{
+        code: "invalid",
+        severity: "error"
+      }])
+    }
+  }
+
+  private static handleErrorOrRejectionResponse(errorCodes: Array<hl7V3.Code<string>>, logger: pino.Logger) {
+    const issues = errorCodes.map(SpineResponseHandler.toOperationOutcomeIssue)
+    if (!issues.length) {
+      logger.error("Trying to return bad request response with no error details")
+      return SpineResponseHandler.createServerErrorResponse()
+    }
+    return SpineResponseHandler.createBadRequestResponse(issues)
+  }
+
+  private static toOperationOutcomeIssue(code: hl7V3.Code<string>): fhir.OperationOutcomeIssue {
+    return {
+      code: "invalid",
+      severity: "error",
+      details: {
+        coding: [{
+          system: "https://fhir.nhs.uk/CodeSystem/Spine-ErrorOrWarningCode",
+          code: code._attributes.code,
+          display: code._attributes.displayName
+        }]
+      }
+    }
+  }
+
+  /* eslint-disable @typescript-eslint/no-unused-vars */
+
+  protected handleRejectionResponse(
+    sendMessagePayload: hl7V3.SendMessagePayload<T>,
+    logger: pino.Logger
+  ): TranslatedSpineResponse {
+    const errorCodes = this.extractRejectionCodes(sendMessagePayload)
+    return SpineResponseHandler.handleErrorOrRejectionResponse(errorCodes, logger)
+  }
+
+  protected handleErrorResponse(
+    sendMessagePayload: hl7V3.SendMessagePayload<T>,
+    logger: pino.Logger
+  ): TranslatedSpineResponse {
+    const errorCodes = this.extractErrorCodes(sendMessagePayload)
+    return SpineResponseHandler.handleErrorOrRejectionResponse(errorCodes, logger)
+  }
+
+  protected handleSuccessResponse(
+    sendMessagePayload: hl7V3.SendMessagePayload<T>,
+    logger: pino.Logger
```

Comment:
doesnt seem like the logger is used here or in any of the extension classes

Reply:
Future proofing (/ over engineering) for when we want to log stuff inside our translation code

Reply:
lol okay :+1: